File: src/main/java/dev/nelmin/ndcore/builders/PotionEffectBuilder.java
package dev.nelmin.ndcore.builders;

import org.bukkit.potion.PotionEffect;
import org.bukkit.potion.PotionEffectType;
import org.jetbrains.annotations.NotNull;

import java.util.Objects;

/**
 * A builder class to simplify the creation and configuration of {@link PotionEffect} objects
 * for use in Bukkit-based plugins. This class allows constructing potion effects with
 * customizable properties such as type, duration, amplifier, visibility of particles, and more.
 * <p>
 * Each method in this class modifies specific properties of the potion effect and supports
 * method chaining by returning the current instance of {@code PotionEffectBuilder}.
 * <p>
 * The {@code build()} method returns the constructed {@link PotionEffect} instance.
 */
public class PotionEffectBuilder {
    private @NotNull PotionEffectType type;
    private int duration = 20; // Default duration (1 second)
    private int amplifier = 0; // Default amplifier (level 1)
    private boolean ambient = false;
    private boolean particles = true;
    private boolean icon = true;

    /**
     * Creates a new PotionEffectBuilder with the specified effect type.
     *
     * @param type the potion effect type
     * @throws NullPointerException if type is null
     */
    public PotionEffectBuilder(@NotNull PotionEffectType type) {
        this.type = Objects.requireNonNull(type, "type cannot be null");
    }

    /**
     * Creates a new PotionEffectBuilder with a default GLOWING effect type.
     */
    public PotionEffectBuilder() {
        this(PotionEffectType.GLOWING);
    }

    /**
     * Sets the effect type of the potion effect.
     *
     * @param type the potion effect type to set
     * @return this builder instance
     * @throws NullPointerException if type is null
     */
    public PotionEffectBuilder effect(@NotNull PotionEffectType type) {
        this.type = Objects.requireNonNull(type, "type cannot be null");
        return this;
    }

    /**
     * Creates a new PotionEffectBuilder with the specified effect type.
     *
     * @param type the potion effect type
     * @return a new PotionEffectBuilder instance
     * @throws NullPointerException if type is null
     */
    public static @NotNull PotionEffectBuilder of(@NotNull PotionEffectType type) {
        return new PotionEffectBuilder(type);
    }

    /**
     * Sets the duration in ticks (20 ticks = 1 second)
     *
     * @param ticks the duration in ticks
     * @return this builder instance
     * @throws IllegalArgumentException if ticks is negative
     */
    public PotionEffectBuilder duration(int ticks) {
        if (ticks < 0) {
            throw new IllegalArgumentException("Duration cannot be negative");
        }
        this.duration = ticks;
        return this;
    }

    /**
     * Sets the duration in seconds (automatically converts to ticks)
     *
     * @param seconds the duration in seconds
     * @return this builder instance
     * @throws IllegalArgumentException if seconds is negative
     */
    public PotionEffectBuilder durationSeconds(int seconds) {
        if (seconds < 0) {
            throw new IllegalArgumentException("Duration cannot be negative");
        }
        this.duration = seconds * 20;
        return this;
    }

    /**
     * Sets the amplifier (power level) of the effect
     *
     * @param level the amplifier level
     * @return this builder instance
     * @throws IllegalArgumentException if level is negative
     */
    public PotionEffectBuilder amplifier(int level) {
        if (level < 0) {
            throw new IllegalArgumentException("Amplifier cannot be negative");
        }
        this.amplifier = level;
        return this;
    }

    /**
     * Sets whether the effect is ambient (has more transparent particles)
     *
     * @param ambient true if the effect should be ambient
     * @return this builder instance
     */
    public PotionEffectBuilder ambient(boolean ambient) {
        this.ambient = ambient;
        return this;
    }

    /**
     * Sets whether the effect should show particles
     *
     * @param particles true if particles should be shown
     * @return this builder instance
     */
    public PotionEffectBuilder particles(boolean particles) {
        this.particles = particles;
        return this;
    }

    /**
     * Sets whether the effect should show an icon in the player's inventory
     *
     * @param icon true if the icon should be shown
     * @return this builder instance
     */
    public PotionEffectBuilder icon(boolean icon) {
        this.icon = icon;
        return this;
    }

    /**
     * Builds and returns a new PotionEffect with the configured properties
     *
     * @return a new PotionEffect instance
     */
    public @NotNull PotionEffect build() {
        return new PotionEffect(
                type,
                duration,
                amplifier,
                ambient,
                particles,
                icon
        );
    }
}

File: src/main/kotlin/dev/nelmin/minecraft/builders/PotionEffectBuilder.kt
package dev.nelmin.minecraft.builders

import org.bukkit.potion.PotionEffect
import org.bukkit.potion.PotionEffectType

/**
 * A builder class for creating customizable `PotionEffect` instances in a more readable and flexible manner.
 *
 * This class allows you to specify and configure properties such as duration, amplifier, ambient effect,
 * particle visibility, and the associated status icon before creating the `PotionEffect` object.
 *
 * The default values for the configuration are:
 * - Duration: 20 ticks (1 second)
 * - Amplifier: 0 (the first level of the effect)
 * - Ambient: false (not considered an ambient effect)
 * - Particles: true (particles are visible)
 * - Icon: true (status icon is displayed)
 *
 * @param type The `PotionEffectType` that represents the type of the potion effect.
 */
class PotionEffectBuilder(private val type: PotionEffectType) {
    /**
     * Represents the duration of the potion effect in ticks.
     *
     * The value determines how long the effect will last when applied.
     * Defaults to 20 ticks (equivalent to 1 second in-game time).
     */
    private var duration: Int = 20 // Default duration (1 second)

    /**
     * Represents the amplifier level for the potion effect.
     * The amplifier determines the strength or potency of the potion effect.
     * Default value is 0, which corresponds to the base level of the effect.
     * For example, higher amplifier values will increase the effect's intensity.
     */
    private var amplifier: Int = 0 // Default amplifier (level 1)

    /**
     * Determines whether the potion effect is considered ambient.
     * Ambient effects have a subtle visual effect compared to non-ambient ones.
     */
    private var ambient: Boolean = false

    /**
     * Determines whether the potion effect will display particles when applied.
     * If set to `true`, particles will be shown; otherwise, no particles will be visible.
     */
    private var particles: Boolean = true

    /**
     * Determines whether the potion effect's icon is displayed in the HUD (Heads-Up Display).
     * When set to `true`, the icon will be visible; otherwise, it will be hidden.
     */
    private var icon: Boolean = true

    /**
     * Sets the duration of the potion effect in game ticks (1 second = 20 ticks).
     *
     * @param ticks The duration of the effect in ticks.
     * @return The current instance of PotionEffectBuilder for method chaining.
     */
    fun duration(ticks: Int): PotionEffectBuilder {
        this.duration = ticks
        return this
    }

    /**
     * Sets the duration of the potion effect in seconds.
     *
     * @param seconds The duration of the potion effect in seconds.
     * This will be internally multiplied by 20 to convert seconds to ticks.
     * @return The current instance of [PotionEffectBuilder] for chaining.
     */
    fun durationSeconds(seconds: Int): PotionEffectBuilder {
        this.duration = seconds * 20
        return this
    }

    /**
     * Sets the amplifier (level) of the potion effect.
     *
     * @param level the amplifier level to set, where 0 represents level 1, 1 represents level 2, and so on
     * @return the updated PotionEffectBuilder instance
     */
    fun amplifier(level: Int): PotionEffectBuilder {
        this.amplifier = level
        return this
    }

    /**
     * Sets whether the potion effect is ambient. Ambient potion effects produce fewer visible particles.
     *
     * @param ambient true if the potion effect should be ambient, false otherwise
     * @return the current instance of PotionEffectBuilder for chaining
     */
    fun ambient(ambient: Boolean): PotionEffectBuilder {
        this.ambient = ambient
        return this
    }

    /**
     * Sets whether particles should be visible for the potion effect.
     *
     * @param particles whether particles should be visible (true to show particles, false to hide them)
     * @return the updated PotionEffectBuilder instance
     */
    fun particles(particles: Boolean): PotionEffectBuilder {
        this.particles = particles
        return this
    }

    /**
     * Sets whether the potion effect will have an icon displayed in the HUD.
     *
     * @param icon true to display the icon in the HUD, false otherwise
     * @return the current instance of PotionEffectBuilder for chaining
     */
    fun icon(icon: Boolean): PotionEffectBuilder {
        this.icon = icon
        return this
    }

    /**
     * Builds and returns a new instance of `PotionEffect` based on the configured properties.
     *
     * @return A newly created `PotionEffect` instance configured with the specified type, duration, amplifier,
     * ambient state, particles visibility, and icon visibility.
     */
    fun build(): PotionEffect {
        return PotionEffect(
            type,
            duration,
            amplifier,
            ambient,
            particles,
            icon
        )
    }

    /**
     * Companion object providing utility methods to create instances of PotionEffectBuilder.
     */
    companion object {
        /**
         * Creates a new instance of `PotionEffectBuilder` with the specified potion effect type.
         *
         * @param type The type of potion effect to be applied.
         * @return A new `PotionEffectBuilder` instance for building potion effects.
         */
        @JvmStatic
        fun of(type: PotionEffectType): PotionEffectBuilder {
            return PotionEffectBuilder(type)
        }
    }
}